torch/_inductor/decomposition.py

```diff
@@ -705,3 +706,50 @@ def _softmax_backward_data(grad_output, output, dim, input_dtype):
     if grad_output.dtype != input_dtype:
         grad_input = grad_input.to(input_dtype)
     return grad_input.contiguous()
+
+
+@register_decomposition(aten.index_reduce)
+def index_reduce(
+    self, dim: int, index, src, reduction_type: str, *, include_self: bool = True
+):
+    if reduction_type == "mean" and not needs_fallback_due_to_atomic_add_limitations(
+        self.dtype
+    ):
+        tmp = (torch.ones_like if include_self else torch.zeros_like)(self)
+        counts = tmp.index_add(dim, index, torch.ones_like(src))
+        out = self if include_self else tmp  # tmp is torch.zeros_like(self)
+        out = out.index_add(dim, index, src)
+        mask = counts == 0
+        if self.dtype.is_floating_point:
+            out = out / counts
+        else:
+            out = out // counts.masked_fill(mask, 1)
```

Comment:
Why do you only include the masked fill in the non-fp case? The C++ version has it outside of the if, so it applies to both cases.

Reply:
First, `mask = counts == 0` is needed only when `include_self==False` (in the `include_self==True` case, minimal count is always 1 or greater).

In the fp-case, there is no need for the `masked_fill` op because true-division (`/`) always succeeds. The `out` entries that have zero counts will be replaced by the corresponding `self` entries (see `where` expression below) and the results from true-zero-division events will be discarded anyway. In this regard, the C++ version is sub-optimal.

In the non-fp-case, floor division (`//`) would throw when dividing by zero, hence the need for the `masked_fill` to protect from floor-zero-division events.

Reply:
> First, mask = counts == 0 is needed only when include_self==False (in the include_self==True case, minimal count is always 1 or greater).

That makes sense, but in that case why mask for `include_self=True` on non-fp? Instead you could do:
```python
if not include_self:
    counts = counts.masked_fill(count == 0, 1)

return out / counts if self.dtype.is_floating_point else out // counts
```

Reply:
The suggested code has two problems:
- it is incomplete as it lacks initializing `out` with `self` entries when `counts==0` is True and `include_self=False`
- it is less optimal as calling `masked_fill` is superfluous when `self.dtype.is_floating_point` is True

Reply:
> it is less optimal as calling masked_fill is superfluous when self.dtype.is_floating_point is True

The `where` in your code is only needed because you don't `masked_fill` for float.